include/pinocchio/algorithm/jacobian.hxx

```diff
@@ -184,50 +185,69 @@ namespace pinocchio
       assert(model.check(data) && "data is not consistent with model.");
 
       PINOCCHIO_CHECK_ARGUMENT_SIZE(Jin.rows(), 6);
-      PINOCCHIO_CHECK_ARGUMENT_SIZE(Jin.cols(), model.nv);
+      PINOCCHIO_CHECK_ARGUMENT_SIZE(Jin.cols(), model.nj);
 
       PINOCCHIO_CHECK_ARGUMENT_SIZE(Jout.rows(), 6);
       PINOCCHIO_CHECK_ARGUMENT_SIZE(Jout.cols(), model.nv);
 
       Matrix6xLikeOut & Jout_ = Jout.const_cast_derived();
+      Jout_.setZero();
 
       typedef typename Matrix6xLikeIn::ConstColXpr ConstColXprIn;
       typedef const MotionRef<ConstColXprIn> MotionIn;
 
       typedef typename Matrix6xLikeOut::ColXpr ColXprOut;
       typedef MotionRef<ColXprOut> MotionOut;
 
-      const int colRef = nv(model.joints[joint_id]) + idx_v(model.joints[joint_id]) - 1;
+      const int colRefIn = nj(model.joints[joint_id]) + idx_j(model.joints[joint_id]) - 1;
       switch (rf)
       {
       case WORLD: {
-        for (Eigen::DenseIndex j = colRef; j >= 0; j = data.parents_fromRow[(size_t)j])
+        for (JointIndex j = joint_id; j > 0; j = model.parents[(size_t)j])
         {
-          MotionIn v_in(Jin.col(j));
-          MotionOut v_out(Jout_.col(j));
+          for (int i = nj(model.joints[j]) - 1; i >= 0; i--)
```

Comment:
Each `nj` call will call `boost::apply_visitor`.  Same for `idx_j` and `idx_v`.

The old algorithm was calling this visitor once, now we call it 3N time.
It's better to use the Algo::run pattern. At least, we will call the visitor only N time.

The only way to not call it at all is:
 - Store idx_j, idx_v and nj in JointModelTpl
 - Create a JacobianComputation structure that will store these values for each joint
 - Store in model a list of support J column for each joint